buildtools/maven/maven.go

```diff
@@ -59,27 +60,75 @@ func (m *Maven) Compile(dir string) error {
 	return err
 }
 
-func Modules(dir string) ([]string, error) {
-	var pom Manifest
-	err := files.ReadXML(&pom, filepath.Join(dir, "pom.xml"))
+// A MvnModule can identify a Maven project with a groupId:artifactId string, a nested
+type MvnModule struct {
+	// Name is taken from the module's POM file.
+	Name string
+
+	// Target is the relative path from the root of the FOSSA project to the module's manifest file or to
+	// the directory containing the module's "pom.xml". The target may name a file other than "pom.xml" and
+	// it may even be the groupId:artifactId string identifying the "Maven project".
+	Target string
+
+	// Path is the relative path from the root of the FOSSA project to the module.
+	Dir string
+
+	GroupId    string
+	ArtifactId string
+}
+
+// Modules returns a list specifying the Maven module at path, which may name a file or directory, and all the
+// Maven modules nested below it. The Target field of each MvnModule is set to the directory or manifest file
+// describing the module. The visited manifest files are listed in the checked map.
+func Modules(path string, checked map[string]bool) ([]MvnModule, error) {
+	cwd, err := os.Getwd()
 	if err != nil {
-		return nil, errors.Wrap(err, "could not read POM file")
+		return nil, errors.Wrap(err, "could not get CWD")
 	}
 
-	var modules []string
-	for _, module := range pom.Modules {
-		children, err := Modules(filepath.Join(dir, module))
+	dir := path
+	pomFile := path
+	if strings.HasSuffix(path, ".xml") {
+		// We have the manifest file but still need its directory path.
+		dir, err = filepath.Rel(cwd, filepath.Dir(path))
 		if err != nil {
-			return nil, errors.Wrap(err, "could not read child modules")
+			return nil, errors.Wrapf(err, "could not get relative path to %q", path)
 		}
-		modules = append(modules, children...)
+	} else {
+		// We have the directory and will assume it uses the standard name for the manifest file.
+		pomFile = filepath.Join(path, "pom.xml")
 	}
 
-	groupID := pom.GroupID
-	if groupID == "" {
-		groupID = pom.Parent.GroupID
+	if checked[pomFile] {
+		return nil, nil
+	}
+
+	checked[pomFile] = true
+
+	var pom Manifest
+	if err := files.ReadXML(&pom, pomFile); err != nil {
+		return nil, errors.Wrapf(err, "could not read POM file %q", pomFile)
+	}
+
+	groupId := pom.GroupID
+	if groupId == "" {
+		groupId = pom.Parent.GroupID
+	}
+
+	modules := make([]MvnModule, 1, 1+len(pom.Modules))
```

Comment:
[nit] could do an inline initialization with something like `modules := []MvnModule{...}`. Also: no need to specify size/capacity

Reply:
no need indeed, but with just one element in a slice literal the slice will be allocated just for one element. The generated code effectively would have `make([]MvnModule, 1, 1)` and then an insertion of the first element. As soon as you add anything, this would mean another allocation and a copy of the first element.

Reply:
Makes sense 👍